<h1 style=">Universidad Castro Carazo</h1>
<h2>Analisis Datos</h2>


<h1 style="color:#a9da88; text-align: center"> Universidad Castro Carazo</h1>
<h2 style="color:#fb7968; text-align: center">Analisis datos</h2>
<br/>
<p style="font-family: sans-serif; text-align:justify">El analisis de datos es un proceso sistematico y metodico que implica la recopilacion e interpretacion y visualizacion de conjuntos de datos con el objetivo de extraer informacion significativa y relevante. Este proceso utiliza herramientas, tecnologias y  procesos para encontrar tendencias y resolver prolemas mediante datos.</p>

<table>
    <tr>
        <td>Repositorio del set de datos</td>
        <td><a href="https://paletadecolores.com.mx/">Enlace</a></td>
    </tr>
    <tr>
        <td>Paleta de colores</td>
        <td><a href="https://paletadecolores.com.mx/">Enlace</a></td>
    </tr>
</table>

https://archive.ics.uci.edu/dataset/10/automobile

<h2>1. Carga y estandarizacion del set de datos</h2>
<ul>
    <li>Instalar las librerias necesarias</li>
    <li>Importar los archivos csv</li>
    <li>Revision general de los set de datos</li>
    <li>Generar un unico Data set</li>
</ul>

In [1]:
import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns

import warnings
warnings.filterwarnings("ignore")

In [2]:
cols = [
    "symboling",
     "Normalized",
    "make",
    "fuel",
    "aspiration",
    "doors",
    "style",
    "wheels",
    "engine",
    "base",
    "length",
    "width",
    "height",
    "curb",
     "engineType",
    "Cylind3rs",
    "engine size",
    "system",
    "bor3",
    "stroke",
    "compression",
    "hOrsepower",
    "peak",
    "CITY",
    "Higway",
    "price"
]

In [3]:
dfAutos1 = pd.read_csv("DataSet_Autos01 4.csv", names=cols)
dfAutos2 = pd.read_csv("DataSet_Autos02 4.csv", names=cols)
print(f"La dimension del set de datos 1 es: {dfAutos1.shape}")
print(f"La dimension del set de datos 2 es: {dfAutos2.shape}")

La dimension del set de datos 1 es: (201, 26)
La dimension del set de datos 2 es: (201, 26)


In [4]:
dfAutos1.head(3)

symboling Normalized         make fuel aspiration doors        style  \
NaN            3          ?  alfa-romero  gas        std   two  convertible   
26.0           1        148        dodge  gas      turbo     ?        sedan   
185.0          2         94   volkswagen  gas        std  four        sedan   

      wheels engine  base  ...  engine size  system  bor3  stroke compression  \
NaN      rwd  front  88.6  ...          130    mpfi  3.47    2.68         9.0   
26.0     fwd  front  93.7  ...           98    mpfi  3.03    3.39         7.6   
185.0    fwd  front  97.3  ...          109    mpfi  3.19    3.40         9.0   

      hOrsepower  peak CITY Higway  price  
NaN          111  5000   21     27  13495  
26.0         102  5500   24     30   8558  
185.0         85  5250   27     34   8495  

[3 rows x 26 columns]

In [5]:
dfAutos2.tail(3)

symboling Normalized    make fuel aspiration doors  style wheels  \
164.0          1        168  toyota  gas        std   two  sedan    rwd   
3.0            2        164    audi  gas        std  four  sedan    4wd   
112.0          0          ?  peugot  gas        std  four  wagon    rwd   

      engine   base  ...  engine size  system  bor3  stroke compression  \
164.0  front   94.5  ...           98    mpfi  3.24    3.08         9.4   
3.0    front   99.4  ...          136    mpfi  3.19    3.40         8.0   
112.0  front  114.2  ...          120    mpfi  3.46    2.19         8.4   

      hOrsepower  peak CITY Higway  price  
164.0        112  6600   26     29   9298  
3.0          115  5500   18     22  17450  
112.0         95  5000   19     24  16695  

[3 rows x 26 columns]

<h4 style="color:#fb7968">1.3) Revisar los encabezados de los set de datos</h4>
<h5>
    Sanas practicas para nombrar las variables:
    <ol>
        <li>Los nombres de variables no deben de empezar con un numero</li>
        <li>Los nombres no pueden ser iguales a una palabra reservada (for, if, print)</li>
        <li>
            Los nombres no deben de contener espacios (engine size -> engineSize / engine-size)
        </li>
        <li>
            Los nombres de variables deben estar en el mismo idioma
        </li>
        <li>
            Los nombres de variables deben estar todos en Minuscula o Mayuscula
        </li>
    </ol>
</h5>

In [6]:
print(f"Listado de variables dfAutos1: {dfAutos1.columns}")
print(f"Listado de variables dfAutos2: {dfAutos2.columns}")

Listado de variables dfAutos1: Index(['symboling', 'Normalized', 'make', 'fuel', 'aspiration', 'doors',
       'style', 'wheels', 'engine', 'base', 'length', 'width', 'height',
       'curb', 'engineType', 'Cylind3rs', 'engine size', 'system', 'bor3',
       'stroke', 'compression', 'hOrsepower', 'peak', 'CITY', 'Higway',
       'price'],
      dtype='object')
Listado de variables dfAutos2: Index(['symboling', 'Normalized', 'make', 'fuel', 'aspiration', 'doors',
       'style', 'wheels', 'engine', 'base', 'length', 'width', 'height',
       'curb', 'engineType', 'Cylind3rs', 'engine size', 'system', 'bor3',
       'stroke', 'compression', 'hOrsepower', 'peak', 'CITY', 'Higway',
       'price'],
      dtype='object')


<h4 style="color:#fb7968">1.4) Generar un unico set de datos</h4>
<h5>
    Puntos a considerar:
    <ol>
        <li>Las variables se llamen igual</li>
        <li>Contener la misma cantidad de variables y estar en el mismo orden</li>
    </ol>
</h5>

In [7]:
unionArchivos = (dfAutos1, dfAutos2)
dfAutos = pd.concat(unionArchivos)
dfAutos.shape

(402, 26)

In [8]:
dfAutos.head(5)

symboling Normalized         make fuel aspiration doors        style  \
NaN            3          ?  alfa-romero  gas        std   two  convertible   
26.0           1        148        dodge  gas      turbo     ?        sedan   
185.0          2         94   volkswagen  gas        std  four        sedan   
112.0          0          ?       peugot  gas        std  four        wagon   
78.0           1        161   mitsubishi  gas      turbo   two    hatchback   

      wheels engine   base  ...  engine size  system  bor3  stroke  \
NaN      rwd  front   88.6  ...          130    mpfi  3.47    2.68   
26.0     fwd  front   93.7  ...           98    mpfi  3.03    3.39   
185.0    fwd  front   97.3  ...          109    mpfi  3.19    3.40   
112.0    rwd  front  114.2  ...          120    mpfi  3.46    2.19   
78.0     fwd  front   93.0  ...           98    spdi  3.03    3.39   

      compression hOrsepower  peak CITY Higway  price  
NaN           9.0        111  5000   21     27  13495  
26.0          7.6        102  5500   24     30   8558  
185.0         9.0         85  5250   27     34   8495  
112.0         8.4         95  5000   19     24  16695  
78.0          7.6        102  5500   24     30   7689  

[5 rows x 26 columns]

In [9]:
dfAutos.reset_index(drop=True, inplace=True)

In [10]:
dfAutos.head(5)

symboling Normalized         make fuel aspiration doors        style  \
0          3          ?  alfa-romero  gas        std   two  convertible   
1          1        148        dodge  gas      turbo     ?        sedan   
2          2         94   volkswagen  gas        std  four        sedan   
3          0          ?       peugot  gas        std  four        wagon   
4          1        161   mitsubishi  gas      turbo   two    hatchback   

  wheels engine   base  ...  engine size  system  bor3  stroke compression  \
0    rwd  front   88.6  ...          130    mpfi  3.47    2.68         9.0   
1    fwd  front   93.7  ...           98    mpfi  3.03    3.39         7.6   
2    fwd  front   97.3  ...          109    mpfi  3.19    3.40         9.0   
3    rwd  front  114.2  ...          120    mpfi  3.46    2.19         8.4   
4    fwd  front   93.0  ...           98    spdi  3.03    3.39         7.6   

  hOrsepower  peak CITY Higway  price  
0        111  5000   21     27  13495  
1        102  5500   24     30   8558  
2         85  5250   27     34   8495  
3         95  5000   19     24  16695  
4        102  5500   24     30   7689  

[5 rows x 26 columns]

In [11]:
dfAutos.tail(5)

symboling Normalized    make fuel aspiration doors      style wheels  \
397          0        106   honda  gas        std   two  hatchback    fwd   
398          0        115   mazda  gas        std  four      sedan    fwd   
399          1        168  toyota  gas        std   two      sedan    rwd   
400          2        164    audi  gas        std  four      sedan    4wd   
401          0          ?  peugot  gas        std  four      wagon    rwd   

    engine   base  ...  engine size  system  bor3  stroke compression  \
397  front   96.5  ...          110    1bbl  3.15    3.58         9.0   
398  front   98.8  ...          122    2bbl  3.39    3.39         8.6   
399  front   94.5  ...           98    mpfi  3.24    3.08         9.4   
400  front   99.4  ...          136    mpfi  3.19    3.40         8.0   
401  front  114.2  ...          120    mpfi  3.46    2.19         8.4   

    hOrsepower  peak CITY Higway  price  
397         86  5800   27     33   7895  
398         84  4800   26     32   8495  
399        112  6600   26     29   9298  
400        115  5500   18     22  17450  
401         95  5000   19     24  16695  

[5 rows x 26 columns]

In [12]:
dfAutos.columns

Index(['symboling', 'Normalized', 'make', 'fuel', 'aspiration', 'doors',
       'style', 'wheels', 'engine', 'base', 'length', 'width', 'height',
       'curb', 'engineType', 'Cylind3rs', 'engine size', 'system', 'bor3',
       'stroke', 'compression', 'hOrsepower', 'peak', 'CITY', 'Higway',
       'price'],
      dtype='object')

In [13]:
dfAutos.rename({'Normalized':'normalized'
                ,'Cylind3rs':'cylinders'
                ,'engine size':'engineSize'
                ,'bor3':'bore'
                ,'hOrsepower':'horsepower'
                ,'CITY':'city'
                ,'Higway':'higway'
               }, axis=1, inplace=True)
#axis 1 es horizontal y axis 0 es vertical

dfAutos.columns

Index(['symboling', 'normalized', 'make', 'fuel', 'aspiration', 'doors',
       'style', 'wheels', 'engine', 'base', 'length', 'width', 'height',
       'curb', 'engineType', 'cylinders', 'engineSize', 'system', 'bore',
       'stroke', 'compression', 'horsepower', 'peak', 'city', 'higway',
       'price'],
      dtype='object')

In [14]:
dfAutos.head(3)

symboling normalized         make fuel aspiration doors        style  \
0          3          ?  alfa-romero  gas        std   two  convertible   
1          1        148        dodge  gas      turbo     ?        sedan   
2          2         94   volkswagen  gas        std  four        sedan   

  wheels engine  base  ...  engineSize  system  bore  stroke compression  \
0    rwd  front  88.6  ...         130    mpfi  3.47    2.68         9.0   
1    fwd  front  93.7  ...          98    mpfi  3.03    3.39         7.6   
2    fwd  front  97.3  ...         109    mpfi  3.19    3.40         9.0   

  horsepower  peak city higway  price  
0        111  5000   21     27  13495  
1        102  5500   24     30   8558  
2         85  5250   27     34   8495  

[3 rows x 26 columns]

<h2>2. Manejo de valores faltantes</h2>
<ol>
    <li>Identificar como se muestran los valores faltantes(NaN, ?, null, EspacioEnBlanco)</li>
    <li>Contar la cantidad de valores faltantes en cada uan de las caracteristicas</li>
    <li>Imputacion simple de valores numericos</li>
    <li>Imputacion simple de valores categoricos</li>
    <li>Eliminar observaciones o registros</li>
</ol>

<h4 style="color:#fb7968">2.1 Identificar como se muestran los valores faltantes</h4>

In [15]:
dfAutos.head(15)

symboling normalized         make fuel aspiration doors        style  \
0           3          ?  alfa-romero  gas        std   two  convertible   
1           1        148        dodge  gas      turbo     ?        sedan   
2           2         94   volkswagen  gas        std  four        sedan   
3           0          ?       peugot  gas        std  four        wagon   
4           1        161   mitsubishi  gas      turbo   two    hatchback   
5           1        118        dodge  gas        std   two    hatchback   
6           3          ?      porsche  gas        std   two      hardtop   
7           1        168       toyota  gas        std   two        sedan   
8           2        104         saab  gas        std  four        sedan   
9           1        148        dodge  gas        std  four    hatchback   
10         -1         74        volvo  gas        std  four        wagon   
11          2        134       toyota  gas        std   two      hardtop   
12          2        164         audi  gas        std  four        sedan   
13          2         94   volkswagen  gas        std  four        sedan   
14          0        161       peugot  gas        std  four        sedan   

   wheels engine   base  ...  engineSize  system  bore  stroke compression  \
0     rwd  front   88.6  ...         130    mpfi  3.47    2.68        9.00   
1     fwd  front   93.7  ...          98    mpfi  3.03    3.39        7.60   
2     fwd  front   97.3  ...         109    mpfi  3.19    3.40        9.00   
3     rwd  front  114.2  ...         120    mpfi  3.46    2.19        8.40   
4     fwd  front   93.0  ...          98    spdi  3.03    3.39        7.60   
5     fwd  front   93.7  ...          90    2bbl  2.97    3.23        9.41   
6     rwd   rear   89.5  ...         194    mpfi  3.74    2.90        9.50   
7     rwd  front   94.5  ...          98    mpfi  3.24    3.08        9.40   
8     fwd  front   99.1  ...         121    mpfi  3.54    3.07        9.30   
9     fwd  front   93.7  ...          90    2bbl  2.97    3.23        9.40   
10    rwd  front  104.3  ...         141    mpfi  3.78    3.15        9.50   
11    rwd  front   98.4  ...         146    mpfi  3.62    3.50        9.30   
12    4wd  front   99.4  ...         136    mpfi  3.19    3.40        8.00   
13    fwd  front   97.3  ...         109    mpfi  3.19    3.40        9.00   
14    rwd  front  107.9  ...         120    mpfi  3.46    3.19        8.40   

   horsepower  peak city higway  price  
0         111  5000   21     27  13495  
1         102  5500   24     30   8558  
2          85  5250   27     34   8495  
3          95  5000   19     24  16695  
4         102  5500   24     30   7689  
5          68  5500   37     41   5572  
6         207  5900   17     25  32528  
7         112  6600   26     29   9298  
8         110  5250   21     28  12170  
9          68  5500   31     38   6229  
10        114  5400   23     28  13415  
11        116  4800   24     30  11199  
12        115  5500   18     22  17450  
13         85  5250   27     34   8195  
14         97  5000   19     24  16630  

[15 rows x 26 columns]

In [16]:
dfAutos["normalized"].unique()

array(['?', '148', '94', '161', '118', '168', '104', '74', '134', '164',
       '90', '115', '158', '119', '85', '93', '150', '101', '231', '65',
       '125', '128', '81', '197', '91', '129', '89', '256', '83', '103',
       '188', '154', '110', '108', '87', '107', '137', '145', '153',
       '102', '194', '122', '77', '106', '142', '113', '192', '121'],
      dtype=object)

In [17]:
dfAutos["stroke"].unique()

array(['2.68', '3.39', '3.40', '2.19', '3.23', '2.90', '3.08', '3.07',
       '3.15', '3.50', '3.19', '3.86', '3.35', '3.58', '3.64', '?',
       '3.41', '3.27', '3.54', '3.46', '3.29', '3.11', '2.64', '3.03',
       '3.52', '3.90', '4.17', '3.16', '2.07', '3.47', '2.76', '3.12',
       '3.10', '2.80'], dtype=object)

In [18]:
dfAutos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 402 entries, 0 to 401
Data columns (total 26 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   symboling    402 non-null    int64  
 1   normalized   402 non-null    object 
 2   make         402 non-null    object 
 3   fuel         402 non-null    object 
 4   aspiration   402 non-null    object 
 5   doors        402 non-null    object 
 6   style        402 non-null    object 
 7   wheels       402 non-null    object 
 8   engine       402 non-null    object 
 9   base         402 non-null    float64
 10  length       402 non-null    float64
 11  width        402 non-null    float64
 12  height       402 non-null    float64
 13  curb         402 non-null    int64  
 14  engineType   402 non-null    object 
 15  cylinders    402 non-null    object 
 16  engineSize   402 non-null    int64  
 17  system       402 non-null    object 
 18  bore         402 non-null    object 
 19  stroke  

<h4 style="color:#fb7968">2.2 Contar los valores faltantes</h4>

In [19]:
dfAutos.replace('?', np.nan, inplace=True)

In [20]:
dfTemp = dfAutos.isnull()
dfTemp.head(10)

symboling  normalized   make   fuel  aspiration  doors  style  wheels  \
0      False        True  False  False       False  False  False   False   
1      False       False  False  False       False   True  False   False   
2      False       False  False  False       False  False  False   False   
3      False        True  False  False       False  False  False   False   
4      False       False  False  False       False  False  False   False   
5      False       False  False  False       False  False  False   False   
6      False        True  False  False       False  False  False   False   
7      False       False  False  False       False  False  False   False   
8      False       False  False  False       False  False  False   False   
9      False       False  False  False       False  False  False   False   

   engine   base  ...  engineSize  system   bore  stroke  compression  \
0   False  False  ...       False   False  False   False        False   
1   False  False  ...       False   False  False   False        False   
2   False  False  ...       False   False  False   False        False   
3   False  False  ...       False   False  False   False        False   
4   False  False  ...       False   False  False   False        False   
5   False  False  ...       False   False  False   False        False   
6   False  False  ...       False   False  False   False        False   
7   False  False  ...       False   False  False   False        False   
8   False  False  ...       False   False  False   False        False   
9   False  False  ...       False   False  False   False        False   

   horsepower   peak   city  higway  price  
0       False  False  False   False  False  
1       False  False  False   False  False  
2       False  False  False   False  False  
3       False  False  False   False  False  
4       False  False  False   False  False  
5       False  False  False   False  False  
6       False  False  False   False  False  
7       False  False  False   False  False  
8       False  False  False   False  False  
9       False  False  False   False  False  

[10 rows x 26 columns]

In [21]:
dfAutos["fuel"].unique()

array(['gas', 'diesel'], dtype=object)

In [22]:
for variable in dfTemp.columns.values.tolist():
    print(dfTemp[variable].value_counts())
    print('')

symboling
False    402
Name: count, dtype: int64

normalized
False    318
True      84
Name: count, dtype: int64

make
False    402
Name: count, dtype: int64

fuel
False    402
Name: count, dtype: int64

aspiration
False    402
Name: count, dtype: int64

doors
False    399
True       3
Name: count, dtype: int64

style
False    402
Name: count, dtype: int64

wheels
False    402
Name: count, dtype: int64

engine
False    402
Name: count, dtype: int64

base
False    402
Name: count, dtype: int64

length
False    402
Name: count, dtype: int64

width
False    402
Name: count, dtype: int64

height
False    402
Name: count, dtype: int64

curb
False    402
Name: count, dtype: int64

engineType
False    402
Name: count, dtype: int64

cylinders
False    402
Name: count, dtype: int64

engineSize
False    402
Name: count, dtype: int64

system
False    402
Name: count, dtype: int64

bore
False    396
True       6
Name: count, dtype: int64

stroke
False    396
True       6
Name: count, dtype: int64


<h4>Axis 1 => columns</h4>
<h4>Axis 0 => rows</h4>

<h4>Fillna metodo</h4>
<h5>ffill => Usar el valor row previo para llenar el na actual</h5>
<h5>bfill => Usar el valor row despues para llenar el na actual</h5>

<h4 style="color:#fb7968">2.3 Imputacion simple de valores numericos</h4> 
Significa que se usa el promedio

In [23]:
def calcular_promedio_int(variable):
    average = round(dfAutos[variable].astype('float').mean(axis=0))
    return average

def calcular_promedio_float(variable):
    average = round(dfAutos[variable].astype('float').mean(axis=0), 2)
    return average

def imputacion_numerica(variable, valor, dfAutos):
    if variable in dfAutos.columns:
        dfAutos[variable].replace(np.nan, valor, inplace=True)
        print("Imputacion realizada de forma correcta")
    else:
        print(f"la columna {variable} no existe en el set de datos")

In [24]:
["normalized", "bore", "stroke", "horsepower", "peak"]
imputacion_numerica("normalized", calcular_promedio_int("normalized"), dfAutos)
imputacion_numerica("peak", calcular_promedio_int("peak"), dfAutos)
imputacion_numerica("bore", calcular_promedio_float("bore"), dfAutos)
imputacion_numerica("stroke", calcular_promedio_float("stroke"), dfAutos)
imputacion_numerica("horsepower", calcular_promedio_int("horsepower"), dfAutos)
imputacion_numerica("peak", calcular_promedio_int("peak"), dfAutos)

Imputacion realizada de forma correcta
Imputacion realizada de forma correcta
Imputacion realizada de forma correcta
Imputacion realizada de forma correcta
Imputacion realizada de forma correcta
Imputacion realizada de forma correcta


In [25]:
dfAutos["peak"].unique()

array(['5000', '5500', '5250', '5900', '6600', '5400', '4800', '5200',
       '5800', '4350', '6000', '4200', '4250', '4400', '4150', 5124,
       '4750', '4500', '5100'], dtype=object)

<h4 style="color:#fb7968">2.4 Imputacion simple de valores categoricos</h4>
Significa que se usa el que mas aparece

In [26]:
dfAutos["doors"].value_counts().idxmax()

'four'

In [27]:
dfAutos["doors"].unique()

array(['two', nan, 'four'], dtype=object)

In [28]:
dfAutos["doors"].replace(np.nan, 'four', inplace=True)

<h4 style="color:#fb7968">2.5 Eliminar Observaciones</h4>

In [29]:
dfAutos.dropna(subset=['price'], axis=0, inplace=True)

In [30]:
dfAutos.reset_index(drop=True, inplace=True)

In [31]:
dfAutos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 394 entries, 0 to 393
Data columns (total 26 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   symboling    394 non-null    int64  
 1   normalized   394 non-null    object 
 2   make         394 non-null    object 
 3   fuel         394 non-null    object 
 4   aspiration   394 non-null    object 
 5   doors        394 non-null    object 
 6   style        394 non-null    object 
 7   wheels       394 non-null    object 
 8   engine       394 non-null    object 
 9   base         394 non-null    float64
 10  length       394 non-null    float64
 11  width        394 non-null    float64
 12  height       394 non-null    float64
 13  curb         394 non-null    int64  
 14  engineType   394 non-null    object 
 15  cylinders    394 non-null    object 
 16  engineSize   394 non-null    int64  
 17  system       394 non-null    object 
 18  bore         394 non-null    object 
 19  stroke  

In [32]:
dfAutos[["bore", "stroke"]] = dfAutos[["bore", "stroke"]].astype(float)
dfAutos[["price", "horsepower", "peak", "normalized"]] = dfAutos[["price", "horsepower", "peak", "normalized"]].astype(int)

dfAutos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 394 entries, 0 to 393
Data columns (total 26 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   symboling    394 non-null    int64  
 1   normalized   394 non-null    int64  
 2   make         394 non-null    object 
 3   fuel         394 non-null    object 
 4   aspiration   394 non-null    object 
 5   doors        394 non-null    object 
 6   style        394 non-null    object 
 7   wheels       394 non-null    object 
 8   engine       394 non-null    object 
 9   base         394 non-null    float64
 10  length       394 non-null    float64
 11  width        394 non-null    float64
 12  height       394 non-null    float64
 13  curb         394 non-null    int64  
 14  engineType   394 non-null    object 
 15  cylinders    394 non-null    object 
 16  engineSize   394 non-null    int64  
 17  system       394 non-null    object 
 18  bore         394 non-null    float64
 19  stroke  

<h2>3. Transformacion de datos</h2>
<ol>
 <li>Formato de los datos</li>
 <li>Tropicalizar datos</li>
<li>Agrupamiento de datos</li>
</ol>

<h4 style="color:#fb7968">3.1 Formateo de Datos</h4>